src/HOCON/Impl/HoconPath.cs

```diff
@@ -0,0 +1,135 @@
+                if (this[i] != parentPath[i])
+                    return false;
+            }
+            return true;
+        }
+
+        public override string ToString()
+            => Value;
+
+        internal static HoconPath FromTokens(HoconTokenizerResult tokens)
+        {
+            if(tokens == null)
+                throw new ArgumentNullException(nameof(tokens));
+
+            var result = new List<string>();
+            var sb = new StringBuilder();
+            while (tokens.Current.Type == TokenType.LiteralValue)
+            {
+                switch (tokens.Current.LiteralType)
+                {
+                    case TokenLiteralType.TripleQuotedLiteralValue:
+                        throw HoconParserException.Create(tokens.Current, null, "Triple quoted string could not be used in path expression.");
+
+                    case TokenLiteralType.QuotedLiteralValue:
+                        // Normalize quoted keys, remove the quotes if the key doesn't need them.
+                        //sb.Append(tokens.Current.Value.NeedQuotes() ? $"\"{tokens.Current.Value}\"" : tokens.Current.Value);
+                        sb.Append(tokens.Current.Value);
+                        break;
+
+                    default:
+                        var split = tokens.Current.Value.Split('.');
+                        for(var i = 0; i < split.Length-1; ++i)
+                        {
+                            sb.Append(split[i]);
+                            result.Add(sb.ToString());
+                            sb.Clear();
+                        }
+                        sb.Append(split[split.Length-1]);
+                        break;
+                }
+                tokens.Next();
+            }
+            result.Add(sb.ToString());
+            return new HoconPath(result);
+        }
+
+        public static HoconPath Parse(string path)
+        {
+            if(path == null)
+                throw new ArgumentNullException(nameof(path));
+
+            return FromTokens(new HoconTokenizer(path).Tokenize());
+        }
+
+        public bool Equals(HoconPath other)
+        {
+            if (other is null) return false;
+            if (Count != other.Count) return false;
+
+            for (var i = 0; i < Count; ++i)
+            {
+                if (this[i] != other[i])
+                    return false;
+            }
+            return true;
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj is null) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            if (obj is HoconPath other) 
+                return Equals(other);
+            return false;
+        }
+
+        public override int GetHashCode()
+        {
+            const int seed = 601;
+            const int modifier = 31;
+
+            unchecked
+            {
+                return this.Aggregate(seed, (current, item) => (current * modifier) + item.GetHashCode());
```

Comment:
From what I've seen, `HoconPath` may be used as key in dictionaries - and using aggregates instead of loops is super expensive way for computing hash code.

Reply:
Changed to a loop. Internally, I only use the last key in the path as a key in the HoconObject field Dictionary, but other people may decide to use them as keys in dictionaries, HoconPath *should* be fixed at that time, though I didn't put the effort to make it into an immutable list, will this be a problem in the future?